// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: simcards/v1/simcards.proto

package v1

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on SimCard with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *SimCard) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on SimCard with the rules defined in the
// proto definition for this message. If any rules are violated, the result is
// a list of violation errors wrapped in SimCardMultiError, or nil if none found.
func (m *SimCard) ValidateAll() error {
	return m.validate(true)
}

func (m *SimCard) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for SimcardId

	// no validation rules for Name

	// no validation rules for Description

	if len(errors) > 0 {
		return SimCardMultiError(errors)
	}

	return nil
}

// SimCardMultiError is an error wrapping multiple validation errors returned
// by SimCard.ValidateAll() if the designated constraints aren't met.
type SimCardMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m SimCardMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m SimCardMultiError) AllErrors() []error { return m }

// SimCardValidationError is the validation error returned by SimCard.Validate
// if the designated constraints aren't met.
type SimCardValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e SimCardValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e SimCardValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e SimCardValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e SimCardValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e SimCardValidationError) ErrorName() string { return "SimCardValidationError" }

// Error satisfies the builtin error interface
func (e SimCardValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sSimCard.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = SimCardValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = SimCardValidationError{}

// Validate checks the field values on SimCards with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *SimCards) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on SimCards with the rules defined in
// the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in SimCardsMultiError, or nil
// if none found.
func (m *SimCards) ValidateAll() error {
	return m.validate(true)
}

func (m *SimCards) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	for idx, item := range m.GetSimCards() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, SimCardsValidationError{
						field:  fmt.Sprintf("SimCards[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, SimCardsValidationError{
						field:  fmt.Sprintf("SimCards[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return SimCardsValidationError{
					field:  fmt.Sprintf("SimCards[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return SimCardsMultiError(errors)
	}

	return nil
}

// SimCardsMultiError is an error wrapping multiple validation errors returned
// by SimCards.ValidateAll() if the designated constraints aren't met.
type SimCardsMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m SimCardsMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m SimCardsMultiError) AllErrors() []error { return m }

// SimCardsValidationError is the validation error returned by
// SimCards.Validate if the designated constraints aren't met.
type SimCardsValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e SimCardsValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e SimCardsValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e SimCardsValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e SimCardsValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e SimCardsValidationError) ErrorName() string { return "SimCardsValidationError" }

// Error satisfies the builtin error interface
func (e SimCardsValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sSimCards.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = SimCardsValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = SimCardsValidationError{}
